src/vs/workbench/contrib/replNotebook/browser/repl.contribution.ts

```diff
@@ -0,0 +1,428 @@
+import { NotebookEditorWidget } from 'vs/workbench/contrib/notebook/browser/notebookEditorWidget';
+import { INotebookEditorService } from 'vs/workbench/contrib/notebook/browser/services/notebookEditorService';
+import { KeyCode, KeyMod } from 'vs/base/common/keyCodes';
+import * as icons from 'vs/workbench/contrib/notebook/browser/notebookIcons';
+import { NOTEBOOK_EDITOR_FOCUSED, REPL_NOTEBOOK_IS_ACTIVE_EDITOR } from 'vs/workbench/contrib/notebook/common/notebookContextKeys';
+import { EditOperation } from 'vs/editor/common/core/editOperation';
+import { EditorContextKeys } from 'vs/editor/common/editorContextKeys';
+import { InputFocusedContext } from 'vs/platform/contextkey/common/contextkeys';
+
+type SerializedNotebookEditorData = { resource: URI; preferredResource: URI; viewType: string; options?: NotebookEditorInputOptions };
+class ReplEditorSerializer implements IEditorSerializer {
+	canSerialize(input: EditorInput): boolean {
+		return input.typeId === ReplEditorInput.ID;
+	}
+	serialize(input: EditorInput): string {
+		assertType(input instanceof ReplEditorInput);
+		const data: SerializedNotebookEditorData = {
+			resource: input.resource,
+			preferredResource: input.preferredResource,
+			viewType: input.viewType,
+			options: input.options
+		};
+		return JSON.stringify(data);
+	}
+	deserialize(instantiationService: IInstantiationService, raw: string) {
+		const data = <SerializedNotebookEditorData>parse(raw);
+		if (!data) {
+			return undefined;
+		}
+		const { resource, viewType } = data;
+		if (!data || !URI.isUri(resource) || typeof viewType !== 'string') {
+			return undefined;
+		}
+
+		const input = instantiationService.createInstance(ReplEditorInput, resource);
+		return input;
+	}
+}
+
+Registry.as<IEditorPaneRegistry>(EditorExtensions.EditorPane).registerEditorPane(
+	EditorPaneDescriptor.create(
+		ReplEditor,
+		REPL_EDITOR_ID,
+		'REPL Editor'
+	),
+	[
+		new SyncDescriptor(ReplEditorInput)
+	]
+);
+
+Registry.as<IEditorFactoryRegistry>(EditorExtensions.EditorFactory).registerEditorSerializer(
+	ReplEditorInput.ID,
+	ReplEditorSerializer
+);
+
+export class ReplDocumentContribution extends Disposable implements IWorkbenchContribution {
+
+	static readonly ID = 'workbench.contrib.replDocument';
+
+	constructor(
+		@INotebookService notebookService: INotebookService,
+		@IEditorResolverService editorResolverService: IEditorResolverService,
+		@IEditorService editorService: IEditorService,
+		@INotebookEditorModelResolverService private readonly notebookEditorModelResolverService: INotebookEditorModelResolverService,
+		@IInstantiationService private readonly instantiationService: IInstantiationService
+	) {
+		super();
+
+		const info = notebookService.getContributedNotebookType('repl');
+
+		// We need to register with the notebook service to let it know about this editor contribution
+		if (!info) {
+			this._register(notebookService.registerContributedNotebookType('repl', {
```

Comment:
It's a big strange that we need to register the notebook type in both core and the ipynb extension, would registering in the core be sufficient?